### PR TITLE
Add Nexus APIs to the Temporal Go SDK

### PR DESCRIPTION
**Pull request by bergundy**
Part of the ongoing effort to integrate Nexus RPC into Temporal.

[Rendered](https://github.com/temporalio/proposals/blob/49b1fc6053d139b54b6f030249463715cd0d9570/nexus/sdk-go.md)

### CODE REVIEW

File: nexus/sdk-go.md
Comment (mfateev) [inline nexus/sdk-go.md:212]: I recommend registering multiple Nexus endpoints with the same worker.

Comment (bergundy) [reply nexus/sdk-go.md:212]: I agree we'll want this but should we do it now or start simple? I anticipate this being the common case.

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:86]: Can this param be nil?

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:95]:
```suggestion
	Workflow func(workflow.Context, I) (O, error)
```

I think we may be able to avoid type aliases in the `temporalnexus` package, but understood if we cannot

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:97]:
```suggestion
	GetStartWorkflowOptions func(context.Context, I, nexus.StartOperationOptions) (client.StartWorkflowOptions, error)
```

`GetX` is a bit clearer in Go than `Getter`. You can remove the "Start" and/or the "Workflow" part if you want though.

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:113]: The purpose of this is a bit confusing, can you clarify this is meant to be used in a handler?

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:174]: This doesn't match the signature of `ExecuteWorkflow` put above. Where's the `nexus.StartOperationOptions` parameter?

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:147]: To clarify, the only reason this exists is for people not wanting to make an options object? But we still make them make a full workflow start options callback? Did they really save much? Maybe instead you should have a `NewWorkflowRunOperationOptions` that accepts these fields instead of top-level overloads.

File: nexus/sdk-go.md
Comment (cretz) [reply nexus/sdk-go.md:212]: Can you clarify? Are you saying this should accept a collection instead of just requiring users to call it multiple times?

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:255]:
```suggestion
func ExecuteNexusOperation(ctx Context, service string, operation any, input any, options NexusOperationOptions) NexusOperationFuture
```

I should be able to provide an operation I defined elsewhere

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:40]: I don't think we should over complicate the API with future compat concerns.
I don't even think that `WorkflowQuery` is an option we want here, I'd make a `NewWorkflowQueryOperation` instead if we want to go that direction but in the previous proposal, you were the one advocating for a single `SyncOperation` concept and giving users a client.

I'm open to renaming this to `NewClientOperation` or `NewSyncClientOperation`.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:58]: Yes, operations have a limited pre-defined interface.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:86]: No, you need at least a workflow ID to start a workflow.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:97]: Sure, I'd prefer `GetOptions` then. I liked `Getter` because it follows the same form as `Handler` but I'm not attached to the name.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:113]: Yes.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:147]: Yes, that's the reason. I believe this will be the preferred way to expose workflows as operations so I made this shorthand form.

The reason I don't like the options struct is that you have to explicitly type the type params but yes a `NewWorkflowRunOperationOptions` also solves that. But keep in mind that even `NewWorkflowRunOperationOptions` may require overloads to support handler and (in the future) a result mapper.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:174]: Oh I added that arg last minute and forgot to update the example.
Thanks for pointing it out.
I was debating taking the options from the context instead of forcing users to pass those in but I think the signature with the operation options is a better approach.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:255]: Good callout, thanks!

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:95]: Looks like it's possible. Thanks!

Comment (bergundy) [reply nexus/sdk-go.md]: Fixed.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md]: Fixed.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md]: Fixed.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:115]: Done.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md]: Changed to `GetOptions`.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:214]: Here's a couple of alternative's I've posted in an internal thread:

The most flexible option is to accept a nexus.Handler in WorkerOptions and have the user construct it via:

```go
reg := &nexus.OperationRegistry{}
_ := reg.Register(op1, op2)
handler, _ = reg.NewHandler()
w := worker.New(client, "task-queue", worker.Options{NexusHandler: handler})
```

^ This allows users to customize the handler completely and intercept calls.

The second option is to accept the registry:

```go
reg := &nexus.OperationRegistry{}
_ := reg.Register(op1, op2)
w := worker.New(client, "task-queue", worker.Options{NexusOperationRegistry: reg})

```

Saves a line of code but loses some flexibility and we'll have to expose interceptors in the Go SDK.

With the first option, you get multiple services OOTB, you just need a "router" handler, which we can also add to the Nexus SDK.

Here's some more alternatives for registering multiple services:

```go
w.RegisterNexusService("service-name", reg)
// -- or --
w.RegisterNexusHandler(handler) // handle can contain a service "router"
// -- or --
w.RegisterNexusHandler("service-name", handler)
```

Using the handler concept gives the most flexibility.

Here are some relevant definitions for reference:

https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#Handler
https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#OperationRegistry
https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#Operation
https://pkg.go.dev/github.com/nexus-rpc/sdk-go@v0.0.7/nexus#RegisterableOperation

Comment (mfateev) [reply nexus/sdk-go.md:214]: I like
```w.RegisterNexusService("service-name", reg)```
and
```w.RegisterNexusHandler("service-name", handler)```

Comment (bergundy) [reply nexus/sdk-go.md:214]: So you would want this to be the only option or should we _also_ support registering the individual operations?

Comment (bergundy) [reply nexus/sdk-go.md:214]: Something that I don't like about forcing specifying the service name is that it creates duplication.
You already have to tell the server to map a service to a namespace and task queue using the incoming service registry APIs.
Specifying the service name also complicates the service renaming process.

So I'd suggest having a way to register a handler or operations without forcing specifying the service name but still supporting backing multiple services with a single worker for these (what I assume are) less common use cases.

If we want a single API, the accepting a `Handler` is the most flexible option.

File: nexus/sdk-go.md
Comment (cretz) [reply nexus/sdk-go.md:40]: > I don't think we should over complicate the API with future compat concerns.

I don't think it's overcomplicating. Can ignore most of my comment, but I would expect `NewSyncOperationWithOptions` to exist just like `NewWorkflowRunOperationWithOptions` does. (having said that, I wish there was only 1 call that took options always, but seems you're settled on 2).

File: nexus/sdk-go.md
Comment (cretz) [reply nexus/sdk-go.md:87]: Can you add that this callback must be set by everyone in the docs? Usually people in Go don't expect to have to provide a workflow ID

File: nexus/sdk-go.md
Comment (cretz) [reply nexus/sdk-go.md:149]: > The reason I don't like the options struct is that you have to explicitly type the type params

I don't think this is that bad of a thing in Go and I think we're adding extra methods do to this personal dislike. But I don't have a super strong opinion just so long as we have proper options-struct-based form we can grow on all of these.

File: nexus/sdk-go.md
Comment (cretz) [inline nexus/sdk-go.md:98]: For others reading, the `Handler` term is commonly used in Go in places like `net/http.Server.Handler` and the `GetOptions` type of name is used in places like `crypto/tls.Config.GetCertificate` (even though it may feel inconsistent that it's not `GetHandler` or `OptionsGetter`)

File: nexus/sdk-go.md
Comment (cretz) [reply nexus/sdk-go.md:214]: To clarify, are we discussing accepting multiple different services? Or a single service but multiple different operations?

Comment (bergundy) [reply nexus/sdk-go.md:214]: Multiple services on the same task queue.

Comment (bergundy) [reply nexus/sdk-go.md:214]: I'm currently thinking we either keep `w.RegisterNexusOperation` assuming multiple services on the same task queue is going to be rarely used or drop this API completely to have a single registration method.

The most flexible approach is to allow registering a handler, either with `w.RegisterNexusHandler` or to pass in the handler in worker options.

For a single service registration would look like this:

```go
reg := &nexus.OperationRegistry{}
_ := reg.Register(op1, op2)
handler, _ = reg.NewHandler()
w.RegisterNexusHandler(handler)
```

For multiple services, we'd have:

```go
service1Reg := &nexus.OperationRegistry{}
_ = service1Reg.Register(op1, op2)
service1Handler, _ = reg.NewHandler()

service2Reg := &nexus.OperationRegistry{}
_ = service2Reg.Register(op3, op4)
service2Handler, _ = reg.NewHandler()

handler := temporalnexus.NewMuxHandler()
handler.RegisterService("svc1", service1Handler)
handler.RegisterService("svc2", service2Handler)

w.RegisterNexusHandler(handler)
```

Comment (mfateev) [reply nexus/sdk-go.md:214]: I don't think we want to support registering operations without associating them to a specific service.

**Comment by bergundy** [comment]
Short update: We're going to add a "service" concept that's a grouping of operations and rename the current "service" concept to "endpoint".

This change has to start from the Nexus API spec and Go SDK and then come into this proposal.
I'm putting this one on hold until we have the prerequisites in place.

File: nexus/sdk-go.md
Comment (mfateev) [inline nexus/sdk-go.md:271]: Consider reversing the order to endpoint, service.

**Review by mfateev — APPROVED** [review]

File: nexus/sdk-go.md
Comment (bergundy) [inline nexus/sdk-go.md:271]: Also considering adding context and an options struct to future proof this API.

File: nexus/sdk-go.md
Comment (bergundy) [reply nexus/sdk-go.md:90]: Hmm.. right, they'll need to if they want their operation to be idempotent. It should be a deterministic function from input to ID.

**Review by cretz — APPROVED** [review]
Approving now that registration is solved

**Pull request merged**
